qiskit/circuit/controlflow/for_loop.py

```diff
@@ -0,0 +1,104 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"Circuit operation representing a ``for`` loop."
+
+import warnings
+from typing import Iterable, Optional
+
+from qiskit.circuit.parameter import Parameter
+from qiskit.circuit.exceptions import CircuitError
+from qiskit.circuit.quantumcircuit import QuantumCircuit
+from .control_flow import ControlFlowOp
+
+
+class ForLoopOp(ControlFlowOp):
+    """A circuit operation which repeatedly executes a subcircuit
+    (``body``) parameterized by a parameter ``loop_parameter`` through
+    the set of integer values provided in ``indexset``.
+
+    Parameters:
+        loop_parameter: The placeholder parameterizing ``body`` to which
+            the values from ``indexset`` will be assigned.
+        indexset: A collection of integers to loop over.
+        body: The loop body to be repeatedly executed.
+        label: An optional label for identifying the instruction.
+
+    **Circuit symbol:**
+
+    .. parsed-literal::
+
+             ┌───────────┐
+        q_0: ┤0          ├
+             │           │
+        q_1: ┤1          ├
+             │  for_loop │
+        q_2: ┤2          ├
+             │           │
+        c_0: ╡0          ╞
+             └───────────┘
+
+    """
+
+    def __init__(
+        self,
+        loop_parameter: Parameter,
+        indexset: Iterable[int],
+        body: QuantumCircuit,
+        label: Optional[str] = None,
+    ):
+        num_qubits = body.num_qubits
+        num_clbits = body.num_clbits
+
+        super().__init__(
+            "for_loop", num_qubits, num_clbits, [loop_parameter, indexset, body], label=label
+        )
```

Comment:
We probably want to be careful here in case `indexset` is a consumable iterable.  We likely want to keep it as-is if it's given as a `range` object (for QASM 3 export, we can convert that into a QASM 3 range, which will be nice), but if not, it might be best to immediately consume `indexset` into a `tuple`.  A user could do something pathological like
```python
def for_loop_indices(vars):
    yield vars[0]
    yield vars[1]
    yield from vars[2]

qc.for_loop(param, for_loop_indices([0, 1, (2, 3, 4)]), ...)
```
which would technically fulfil all the type hints and whatnot, but would likely fail the _second_ time something tries to read the indices.

Reply:
Good catch. The choice of `Iterable` was chosen to avoid the single-use problem of accepting an `Iterator`, but it looks like we missed this case ( that if you accept an `Iterable` which was built from an `Iterator`, it is still only single use).

I'd rather not implement immediate consumption, since it adds more state to the `Instruction` that needs to be carried around (and likely would defeat any benefit from providing an `Iterator` in the first place). We do want to support python `range` making it to QASM3 range on export, so maybe `Sequence` is a better choice here?